internal/backend/media.go

```diff
@@ -0,0 +1,99 @@
+package backend
+
+import (
+	"context"
+	"io/ioutil"
+	"mime/multipart"
+	"strings"
+
+	"go.uber.org/multierr"
+
+	"github.com/nedrocks/delphisbe/graph/model"
+
+	"github.com/nedrocks/delphisbe/internal/util"
+
+	"github.com/sirupsen/logrus"
+)
+
+func (d *delphisBackend) GetMedia(ctx context.Context, mediaID string, mediaType string) ([]byte, error) {
+	return d.mediadb.GetMedia(ctx, mediaID, mediaType)
+}
+
+func (d *delphisBackend) UploadMedia(ctx context.Context, ext string, media multipart.File) (string, string, error) {
+	uuid := util.UUIDv4()
+	filename := strings.Join([]string{uuid, ext}, "")
+
+	mediaBytes, err := ioutil.ReadAll(media)
+	if err != nil {
+		logrus.WithError(err).Error("failed to read all media bytes")
+		return "", "", err
+	}
+
+	// Pass in size into s3
+	mimeType, err := d.mediadb.UploadMedia(ctx, filename, mediaBytes)
+	if err != nil {
+		logrus.WithError(err).Error("failed to upload media to s3")
+		return "", "", err
+	}
+
+	mediaSize := getMediaSize(ctx, mimeType, mediaBytes)
+
+	// Create record within Media table
+	mediaObj := model.Media{
+		ID:   uuid,
+		Type: mimeType,
+		Size: &mediaSize,
+	}
+
+	if err := d.writeMediaRecord(ctx, mediaObj); err != nil {
+		logrus.WithError(err).Error("failed to put media record in db")
```

Comment:
creating the media record greatly simplifies the client business logic. 
- UploadImage uploads the media to s3 and creates a record within the media table
- AppPost creates a record in the post table with the foreign key pointing to the record in the media table

if we write the media record within AddPost, the client will have to keep track of ID, size, mimetype, etc.

Reply:
I :100: agree with you here. The two phase creation of media objects is a solid approach.